Week-10/exercise-10.1.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * Class for creating a new tree node
+ */
+class Node {
+  constructor(val, left = null, right = null) {
+    this.value = val;
+    this.left = left;
+    this.right = right;
+  }
+}
+
+/**
+ * Class for creating a new binary search tree
+ */
+class BST {
+  constructor(...items) {
+    this.root = null;
+
+    items?.forEach((item) => this.insertNode(item));
+  }
+
+  /**
+   * Inserts a new node to the binary search tree
+   * !Time Complexity: O(logn)
+   * !Space Complexity: O(1)
+   * @param {Node} root - The root node of the tree where the new node
+   * will be inserted
+   * @param {number} value - The value of the new node that will be inserted
+   */
+  insertNode(value) {
+    const insertNodeHelper = (root, val) => {
+      if (root == null) {
+        return new Node(val);
+      }
+
+      if (root.value < val) {
+        root.right = insertNodeHelper(root.right, val);
+      } else if (root.value > val) {
+        root.left = insertNodeHelper(root.left, val);
+      }
+
+      return root;
+    };
+    this.root = insertNodeHelper(this.root, value);
+  }
+
+  /**
+   * Finds the height of a binary search tree
+   * !Time Complexity: O(n)
+   * !Space Complexity: O(1)
+   * @param {Node} root The root node of the tree from which height will be
+   * calculated
+   * @returns {number} The height of the BST
+   */
+  height(root = this.root) {
+    const heightHelper = (rt) => {
+      if (rt == null) {
+        return 0;
+      }
+      return 1 + Math.max(heightHelper(rt.left), heightHelper(rt.right));
```

Comment:
do you think if there is any limitation to this function?
what max height of BST do u think we can calculate?

Reply:
one edge case could be skewed BSTs (where nodes are only on one side). But the solution works for that case as well.